### PR TITLE
Optimize TupleDomain.columnWiseUnion during dynamic filtering partition collection

### PR DESCRIPTION
**Pull request by rzeyde-varada**
Otherwise, the collection complexity is quadratic in the number of partitions.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 1]
```diff
@@ -72,21 +73,21 @@ public LocalDynamicFilter(Multimap<String, Symbol> probeSymbols, Map<String, Int
...
         // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
         // See the comment at TupleDomain::columnWiseUnion() for more details.
-        result = TupleDomain.columnWiseUnion(result, tupleDomain);
-        if (partitionsLeft == 0) {
+        partitions.add(tupleDomain);
+        if (partitions.size() == partitionCount) {
+            Map<Symbol, Domain> result = convertTupleDomain(TupleDomain.columnWiseUnion(partitions));
             // No more partitions are left to be processed.
-            verify(resultFuture.set(convertTupleDomain(result)), "dynamic filter result is provided more than once");
+            verify(resultFuture.set(result), "dynamic filter result is provided more than once");
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:90]: This `verify` can never fail as there is no way to enter this `if` block multiple times.

Comment (rzeyde-varada) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: Thanks, fixed in d7f7cdd.

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 2]
```diff
@@ -57,11 +58,11 @@
...
 
-    // The resulting predicate for local dynamic filtering.
-    private TupleDomain<String> result;
+    // Number of build-side partitions to be collected.
+    private final int partitionCount;
 
-    // Number of partitions left to be processed.
-    private int partitionsLeft;
+    // The resulting predicates from each build-side partition.
+    private List<TupleDomain<String>> partitions;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:65]: make this `final`

Comment (rzeyde-varada) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: Thanks, fixed in 5895edc03a.

**Comment by rzeyde-varada** [comment]
Many thanks!

**Pull request merged**